---
title: "Bulk RNAseq report (Chiocca Group, project PRESERVE, preserve paper)"
author: 
- Zhan Yinxiu
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  workflowr::wflow_html:
    toc_depth: 4
editor_options:
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo       = FALSE,
                      message    = FALSE,
                      warning    = FALSE,
                      cache      = FALSE,
                      autodep    = TRUE,
                      fig.align  = 'center',
                      fig.width  = 6,
                      fig.height = 6,
                      dpi =        300)
```


```{r libraries}
library("DESeq2")
library(QuasR)
library(rtracklayer)
library(ComplexHeatmap)
library(RColorBrewer)
library(ComplexHeatmap)
library(gplots) 
library(ggplot2)
library("DT")
library(reshape)
library(plotly)
library(htmlwidgets)
library(GSVA)
```


```{r functions}

mydeseq2 = function(dds, min_reads, conditions, min_sample = 5, ...){
  # recreate dds object
  counts = assay(dds)
  counts = counts[rowSums(counts) > min_reads & rowSums(counts > 1) > min_sample, ]
  coldata<-as.data.frame(response[colData(dds)$sample])
  colnames(coldata)="condition"
  coldata$condition = as.factor(coldata$condition)
  rownames(coldata)<-colnames(counts)
  coldata$functionality = functionality[colData(dds)$sample]
  
  
  ##Performing deseq2
  dds_current <-DESeqDataSetFromMatrix(
    countData=counts,
    colData=coldata,
    design=~condition)
  
  diff<-DESeq(dds_current, ...)
  res_all = NULL
  df_all_genes = NULL
  ##Extracting results
  for(i in 1:length(conditions)){
    for(j in 1:length(conditions)){
      if(i!=j){
        res<-results(diff, contrast = c("condition", conditions[i], conditions[j])) # cond1 vs cond2
        res = data.frame(res)
        res$comparison_n_vs_d = paste0(conditions[i], "_VS_", conditions[j])
        res$gene = rownames(res)
        res_all = rbind(res_all, res)
        res = na.omit(res)
        df_all_genes = rbind(df_all_genes, res)
      }
    }
  }
  # normalisation
  rld <- vst(dds_current, blind=FALSE)
  return(list(res=res_all, df_all_genes=df_all_genes, rld=rld, coldata=coldata))
}



make_plots = function(res, column_names){
  res = na.omit(res)
  select = which((res)$padj < qvalue &
                   abs((res)$log2FoldChange) > logfc)

  res = res[select,]
  
  genes = unique(res$gene[res$comparison_n_vs_d %in% contrasts])
  
  counts = assay(rld)
  counts = t(apply(counts, 1, scale))
  colnames(counts) = colnames(assay(rld))
  
  ordering = order(column_names)
  rng <- quantile(counts, c(.025, 0.975))
  column_ha = HeatmapAnnotation(condition = column_names[ordering])
  print(Heatmap(
    matrix = counts[, ordering],
    col = circlize::colorRamp2(seq(rng[1], rng[2], len =
                                     9), brewer.pal(9, "RdBu")),
    name = "Expression across patients",
    cluster_rows = TRUE,
    cluster_columns = FALSE,
    show_column_names = FALSE,
    show_row_names = FALSE,
    width = unit(4, "inches"),
    show_heatmap_legend = TRUE,
    use_raster = FALSE,
    raster_device = "png",
    raster_quality = 1, 
    top_annotation = column_ha,
    km = 3
  ))
  return(res)
}

mydatatable <- function(x, rownames = FALSE){
  DT::datatable(
  x,
  extensions = 'Buttons',
  options = list(
    dom = 'Blfrtip',
    buttons = c('copy', 'csv', 'excel', 'pdf'),
    lengthMenu = list(c(10, 25, 50,-1),
                      c(10, 25, 50, "All"))
  ),
  rownames = rownames
)
}

```


# Objective

RNA-seq data from the Preserve study, generated by Micaela, was analyzed using the [nf-core RNA-seq pipeline](https://nf-co.re/rnaseq) (v3.11.2) with STAR for mapping. This report provides an overview of quality control metrics, differential gene expression analysis results, and corresponding heatmaps.

Below are the parameters used for differential gene expression analysis and filtering criteria:

- q-value threshold: 0.1
- Log fold change (logFC): log₂(1)
- Minimum sample threshold: 3 (a gene must have at least 1 read in a minimum of 3 samples)
- min reads: 

```{r parameters}
rdata_file="/hpcnfs/scratch/DIMA/zhan/chiocca/micaela/230913_rnaseq/results/star_salmon/deseq2_qc/deseq2.dds.RData"
samples_info_file = "/hpcnfs/scratch/DIMA/zhan/chiocca/micaela/230913_rnaseq/samples_files/samples_info.txt"
qvalue=0.1
logfc=log2(1)
min_sample = 3 # minimum number of samples where the gene needs to have at least 1 read
signature = c("PTGS2", "PLAC8", "RNF138", "OGFRL1", "LRRD1", "AL161418.1") # genes in signature from https://academic.oup.com/jnci/article/113/4/471/5872531

signature_hotscore = list("CCR2", "CCL19",  "CCR4", "CCR5", "CD27", "CD40LG", "CD8A", "CXCL10", "CXCL11", "CXCL13", "CXCL9", "CXCR3", "CXCR6", "FASLG", "FGL2", "GZMA", "GZMH", "IDO1", "IFNG", "IRF8", "LAG3", "LYZ", "MS4A1", "PDCD1", "TBX21", "TLR7", "TLR8") # from https://pubmed.ncbi.nlm.nih.gov/36038492/

pattern_remove_gene = "^TRNA" # remove genes strating with TRNA
contrasts = c(
  "COMPLETE.RESPONSE_VS_PROGRESSIVE.DISEASE",
  "COMPLETE.RESPONSE_VS_STABLE.DISEASE",
  "COMPLETE.RESPONSE_VS_PARTIAL.RESPONSE",
  "PARTIAL.RESPONSE_VS_PROGRESSIVE.DISEASE",
  "PARTIAL.RESPONSE_VS_STABLE.DISEASE",
  "PROGRESSIVE.DISEASE_VS_STABLE.DISEASE"
) # comparison to analyse
```


# Quality control

Below is the distribution of the number of uniquely mapped reads per sample. Samples with extremely poor quality—specifically those with fewer than 1 million reads (indicated by the vertical red line)—were excluded from the analysis.

```{r, data_loading}
load(rdata_file)
samples_info = read.delim(samples_info_file, sep="\t", header=T)
samples_info = samples_info[make.names(samples_info$ID.e.CRF) %in% colnames(dds),]
samples_info = samples_info[!is.na(samples_info$Status.of.disease.after.IC),]


```

```{r filtering}
# Remove bad quality reads 
tmp = data.frame(coverage = colSums(assay(dds)))
tmp$quality = "Good"
exclude = rownames(tmp)[tmp$coverage < 1e6]
tmp$quality[rownames(tmp) %in% exclude] = "Bad"
dds = dds[, -grep("_", colnames(dds))]
dds = dds[, !(colnames(dds) %in% exclude)]

# remove samples without the clinical information as they have been excluded
dds = dds[,(colnames(dds) %in% make.names(samples_info$ID.e.CRF))]
dds_tosave = dds

# save raw counts
write.table(x=assay(dds_tosave), file="counts_raw.csv", quote=F, sep=",")


samples_info$path = basename(samples_info$path)
samples_info$resequencing.path = basename(samples_info$resequencing.path)

# assign response to sample file
response = toupper(make.names(samples_info$Status.of.disease.after.IC))
names(response) = make.names(samples_info$ID.e.CRF)

# assign tumor_site to sample file
tumor_site = toupper(make.names(samples_info$Tumor.site))
names(tumor_site) = make.names(samples_info$ID.e.CRF)

# assign center to sample file
center = toupper(make.names(samples_info$Center))
names(center) = make.names(samples_info$ID.e.CRF)

# assign ischeme to sample file
ischeme = toupper(make.names(samples_info$Induction.scheme))
names(ischeme) = make.names(samples_info$ID.e.CRF)

# assign functionality to sample
functionality = make.names(samples_info$LEDFS..0.lar.funzionante..1.disfunzione.)
names(functionality) = make.names(samples_info$ID.e.CRF)

# remove TRNA
pattern_remove_gene = "^TRNA"
keep = rownames(dds)[!grepl(pattern_remove_gene, rownames(dds))]
dds = dds[keep, !(colnames(dds) %in% names(response[response == "X"]))]
response = response[response != "X"]
```

The list of filtered samples is shown below

```{r}
tmp = samples_info[samples_info$ID.e.CRF %in% gsub("\\.", "-", exclude),]
tmp = tmp[,c(1:(ncol(tmp)-2))]
colnames(tmp) = c("sample_id", "tumor_site", "center", "ischeme", "OS", "PFS", "LEDFS", "IC_response", "detailed_status")

mydatatable(tmp)

```

# Differential gene expression analysis

Differential gene expression analysis was performed using  [DESeq2](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)
The following additional criteria was used to define differentially expressed genes: genes were retained for analysis if their total read count across all samples was at least half the total number of samples.

## Resistant vs responsive

The list of contrast is shown below

```{r}
mydatatable(data.frame(contrasts = contrasts))
```

```{r mild_filtering_analysis, include=FALSE}
mreads = round(nrow(samples_info) / 2) ## filter genes without at least mreads reads in all condition combined
out = mydeseq2(dds, min_reads = mreads, min_sample = min_sample, conditions= unique(response), test="Wald")
res = out[["res"]]
rld = out[["rld"]]
df_all_genes = out[["df_all_genes"]]
coldata = out[["coldata"]]
```


Below is the PCA plot, with samples colored based on functionality. While clear separation between functional and non-functional samples is not evident, there is a slight shift toward the right for non-functional samples, suggesting potential trends worth further exploration.

```{r pca_functionality}
res = na.omit(res)

rld$functionality[rld$functionality == "X1"] = "not_functional"
rld$functionality[rld$functionality == "X0"] = "functional"
plotPCA(rld, intgroup = "functionality")
```

The PCA plot below shows samples colored by response status. Unfortunately, there is no clear separation between responders and non-responders, indicating that PCA alone is insufficient to distinguish response groups in this dataset.

```{r pca_response}
plotPCA(rld)
```

Below is the heatmap displaying the z-scores of gene expression. Genes have been clustered based on functionality and response. The heatmap in pdf can be found [here](https://github.com/dimadatascience/preserve/tree/main/output)


```{r}
functionality[functionality == "X0"] = "functional"
functionality[functionality == "X1"] = "not_functional"

response_functionality = response
response_functionality = paste0(response, "_", functionality[names(response_functionality)])
names(response_functionality) = names(response)

res = na.omit(res)
select = which((res)$padj < qvalue &
                 abs((res)$log2FoldChange) > logfc)
res = res[select, ]

genes = unique(res$gene[res$comparison_n_vs_d %in% contrasts])

counts = assay(rld)
counts = t(apply(counts, 1, scale))
colnames(counts) = colnames(assay(rld))

column_names = response_functionality[colnames(assay(rld))]
ordering = order(column_names)
rng <- quantile(counts, c(.025, 0.975))
column_ha = HeatmapAnnotation(condition = column_names[ordering])
```



```{r heatmap_functionality_response, fig.width=10}
Heatmap(
  matrix = counts[, ordering],
  col = circlize::colorRamp2(seq(rng[1], rng[2], len =
                                   9), brewer.pal(9, "RdBu")),
  name = "Expression across patients",
  cluster_rows = TRUE,
  cluster_columns = TRUE,
  show_column_names = FALSE,
  show_row_names = FALSE,
  width = unit(4, "inches"),
  show_heatmap_legend = TRUE,
  use_raster = FALSE,
  raster_device = "png",
  raster_quality = 1,
  top_annotation = column_ha,
  column_split = column_names[ordering],
  km = 3
)
```



```{r include=FALSE}
pdf("/hpcnfs/scratch/DIMA/zhan/chiocca/micaela/preserve/output/heatmap4paper.pdf")
Heatmap(
  matrix = counts[, ordering],
  col = circlize::colorRamp2(seq(rng[1], rng[2], len =
                                   9), brewer.pal(9, "RdBu")),
  name = "Expression across patients",
  cluster_rows = TRUE,
  cluster_columns = TRUE,
  show_column_names = FALSE,
  show_row_names = FALSE,
  width = unit(4, "inches"),
  show_heatmap_legend = TRUE,
  use_raster = FALSE,
  raster_device = "png",
  raster_quality = 1,
  top_annotation = column_ha,
  column_split = column_names[ordering],
  km = 3
)
dev.off()
```

The transcriptomic profiles of responsive and resistant samples appear to be highly similar, with minimal differences observed. Below is a table highlighting the few differentially expressed genes identified between the two groups.


```{r mf_table}
datatable(
  res[(res$comparison_n_vs_d %in% contrasts),],
  extensions = 'Buttons',
  options = list(
    dom = 'Blfrtip',
    buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
    lengthMenu = list(c(10, 25, 50, -1),
                      c(10, 25, 50, "All"))
  ),
  rownames = FALSE
)

```

Next, we focused on the expression changes for the following gene signature:
Signature: c("PTGS2", "PLAC8", "RNF138", "OGFRL1", "LRRD1", "AL161418.1")

It is important to note that the gene AL161418.1 (alias ENST00000640280.1) has been removed in the latest Ensembl transcript annotation. For details, refer to the [link](https://www.ensembl.org/Homo_sapiens/Transcript/Idhistory?t=ENST00000640280)

Below, we present both the table of expression changes and the corresponding boxplot for visualization.

```{r mf_list_genes}
df_filtered_genes = df_all_genes[(df_all_genes$gene %in% signature) & (df_all_genes$comparison_n_vs_d %in% contrasts), ]
datatable(
  df_filtered_genes,
  extensions = 'Buttons',
  options = list(
    dom = 'Blfrtip',
    buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
    lengthMenu = list(c(10, 25, 50, -1),
                      c(10, 25, 50, "All"))
  ),
  rownames = FALSE
)

```


```{r normalised_data}
tmp = assays(dds)[["vst"]]
# tmp = t(t(tmp) / colSums(tmp)) * 1e6
tmp = tmp[rownames(tmp) %in% signature, ]
tmp = melt(tmp)
tmp$response = (response[tmp$X2])
tmp$response1 = "RESISTANT"
tmp$response1[grepl("RESPONSE",tmp$response)] = "RESPONSIVE"

```

```{r signature_boxplot}
ggplot(data = tmp, aes(x=X1, y=(value), fill = response1)) + geom_boxplot() + ylab("Expression value (vst normalised)") + xlab("gene")
```

# Hotscore

Next, we analyzed the hotscore across patients and generated a table detailing the gene expression changes of the hotscore-related genes. This analysis highlights the expression dynamics of these genes across the patient cohort.


```{r hotscore_analysis, results=FALSE}
# define it as single list
signature_hotscore = list(c("CCR2", "CCL19",  "CCR4", "CCR5", "CD27", "CD40LG", "CD8A", "CXCL10", "CXCL11", "CXCL13", "CXCL9", "CXCR3", "CXCR6", "FASLG", "FGL2", "GZMA", "GZMH", "IDO1", "IFNG", "IRF8", "LAG3", "LYZ", "MS4A1", "PDCD1", "TBX21", "TLR7", "TLR8")) # from https://pubmed.ncbi.nlm.nih.gov/36038492/

tmp = assays(dds)[["vst"]]
hotscores = gsva(tmp, signature_hotscore)
# rownames(hotscores) = as.vector(signature_hotscore)

hotscores = data.frame(t(hotscores))
colnames(hotscores) = "hotscore"

a = response[rownames(hotscores)]
names = rep("RESISTANT", length(a))
names[grepl("RESPONSE", a)] = "RESPONSIVE"

# hotscores$condition = names
hotscores$condition = response[rownames(hotscores)]
hotscores$hot = hotscores$hotscore>0

table(hotscores[,c("condition", "hot")])
```

```{r hotscore_plot}

#pdf("hotscore197patients4susanna.pdf")
hotscores$condition = response[rownames(hotscores)]
ggplot(data = hotscores, aes(x=condition, y=hotscore)) + geom_boxplot() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

hotscores$condition[hotscores$condition %in% c("STABLE.DISEASE", "PROGRESSIVE.DISEASE")] = "RESISTANT"
hotscores$condition = response[rownames(hotscores)]
ggplot(data = hotscores, aes(x=condition, y=hotscore)) + geom_boxplot() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
#dev.off()

hotscores$hot = "hot"
hotscores$hot[hotscores$hotscore<0] = "cold"


hotscores$condition = functionality[rownames(hotscores)]
ggplot(data = hotscores, aes(x=condition, y=hotscore)) + geom_boxplot() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + xlab("Functionality")
```


```{r}
myfunc = function(dds, min_reads, conditions, min_sample = 5, ...){
  # recreate dds object
  counts = assay(dds)
  counts = counts[rowSums(counts) > min_reads & rowSums(counts > 1) > min_sample, ]
  coldata<-as.data.frame(response1[colData(dds)$sample])
  colnames(coldata)="condition"
  coldata$condition = as.factor(coldata$condition)
  rownames(coldata)<-colnames(counts)
  #coldata$ischeme = ischeme[colData(dds)$sample]
  #coldata$center = center[colData(dds)$sample]
  #coldata$tumor_site = tumor_site[colData(dds)$sample]
  coldata$functionality = functionality[colData(dds)$sample]
  #coldata$batch = batch[colData(dds)$sample]
  
  
  
  ##Performing deseq2
  dds_current <-DESeqDataSetFromMatrix(
    countData=counts,
    colData=coldata,
    design=~condition)
  
  diff<-DESeq(dds_current, ...)
  res_all = NULL
  df_all_genes = NULL
  ##Extracting results
  for(i in 1:length(conditions)){
    for(j in 1:length(conditions)){
      if(i!=j){
        res<-results(diff, contrast = c("condition", conditions[i], conditions[j])) # cond1 vs cond2
        res = data.frame(res)
        res$comparison_n_vs_d = paste0(conditions[i], "_VS_", conditions[j])
        res$gene = rownames(res)
        res_all = rbind(res_all, res)
        res = na.omit(res)
        df_all_genes = rbind(df_all_genes, res)
      }
    }
  }
  # normalisation
  rld <- vst(dds_current, blind=FALSE)
  return(list(res=res_all, df_all_genes=df_all_genes, rld=rld, coldata=coldata))
}

mreads = round(nrow(samples_info) / 2) ## filter genes without at least mreads reads in all condition combined
response1 = response
response1[response1=="STABLE.DISEASE"] = "SD.PD"
response1[response1=="PROGRESSIVE.DISEASE"] = "SD.PD"
out = myfunc(dds, min_reads = mreads, min_sample = min_sample, conditions= unique(response1), test="Wald")
res = out[["res"]]
rld = out[["rld"]]
df_all_genes = out[["df_all_genes"]]
coldata = out[["coldata"]]
```

```{r}
select = which((res)$padj < 0.05 &
                   abs((res)$log2FoldChange) > log2(2))
#tmp = plotPCA(rld, intgroup = "condition", returnData=TRUE)
#print(plotly::plot_ly(data=tmp, x=~PC1, y=~PC2, text = ~name, color=~condition))
# res = res[select,]
res = res[res$gene %in% unlist(signature_hotscore),]
res = res[res$comparison_n_vs_d %in% c("SD.PD_VS_COMPLETE.RESPONSE", "SD.PD_VS_PARTIAL.RESPONSE"),]
```

```{r}
mydatatable(res, rownames = TRUE)
```
